### PR TITLE
[SPARK-17899][SQL] add a debug mode to keep raw table properties in HiveExternalCatalog

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently `HiveExternalCatalog` will filter out the Spark SQL internal table properties, e.g. `spark.sql.sources.provider`, `spark.sql.sources.schema`, etc. This is reasonable for external users as they don't want to see these internal properties in `DESC TABLE`.

However, as a Spark developer, sometimes we do wanna see the raw table properties. This PR adds a new internal SQL conf, `spark.sql.debug`, to enable debug mode and keep these raw table properties.

This config can also be used in similar places where we wanna retain debug information in the future.
## How was this patch tested?

new test in MetastoreDataSourcesSuite

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by rxin** [comment]
LGTM pending Jenkins.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala [hunk 1]
```diff
@@ -254,18 +255,21 @@ class SQLConfSuite extends QueryTest with SharedSQLContext {
...
-  test("global SQL conf comes from SparkConf") {
-    val newSession = SparkSession.builder()
-      .config(SCHEMA_STRING_LENGTH_THRESHOLD.key, "2000")
-      .getOrCreate()
-
-    assert(newSession.conf.get(SCHEMA_STRING_LENGTH_THRESHOLD.key) == "2000")
-    checkAnswer(
-      newSession.sql(s"SET ${SCHEMA_STRING_LENGTH_THRESHOLD.key}"),
-      Row(SCHEMA_STRING_LENGTH_THRESHOLD.key, "2000"))
+  test("static SQL conf comes from SparkConf") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: this change is not related to this PR.

`SparkSession.builder()....getOrCreate()` may not set the static SQL conf if there is an active or global default session. To make this test more robust, I explicitly create a new SparkSession.

This is the same approach I used in the new test, so I include this unrelated change in this PR.

**Comment by andrewor14** [comment]
Cool beans. Merging into master 2.0.

**Comment by andrewor14** [comment]
JK, actually it doesn't merge in 2.0.

**Comment by gatorsmile** [comment]
If we turn this flag on, I am just afraid that some DDL statements might not work well. I assume this flag is really used for the our Spark SQL developers.

**Comment by andrewor14** [comment]
Yes that's why it's `internal`

**Comment by gatorsmile** [comment]
Just FYI. https://github.com/apache/spark/pull/15478 shows at least three DDL commands do not work properly when `spark.sql.debug` is set to `true`: `ALTER TABLE`, `ANALYZE TABLE` and `CREATE TABLE LIKE`

**Comment by andrewor14** [comment]
I see. Then maybe we should add a comment above the config to note that several commands don't work (e.g. ALTER TABLE) if this is turned on, even if it's only internal.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -915,4 +915,9 @@ object StaticSQLConf {
     .internal()
     .intConf
     .createWithDefault(4000)
+
+  val DEBUG_MODE = buildConf("spark.sql.debug")
+    .internal()
+    .booleanConf
+    .createWithDefault(false)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Sure, I can do a quick PR to fix it.
